Add number type to fontWeight & fontWeightBold options

IllusionMH opened the pull request:
Fixes #3051 

`_sanitizeAndValidateOption` now implements custom normalization for `fontWeight` & `fontWeightBold` options that matches normalization from [VSCode PR](https://github.com/microsoft/vscode/pull/101468/files#diff-bf729abb47c6d125996e13230061c850R348-R353).
This is needed because if invalid value (`7000` or `normal123`) is passes to `fontWeight` when canvas or webgl rendenderer is used it will lead to really small characters.
![image](https://user-images.githubusercontent.com/1312662/91764579-3c97bb80-ebe0-11ea-8ea6-1f37debdb4f6.png)

In browsers that don't support numeric values for `font-weight` property when values that doesn't match existing enum would be reset to `normal`/`400` - this is current browser behavior for `font-weight` and no special logic added to handle this case. I assume that users of these browsers shouldn't won't use numeric values e.g. 350 because they don't work anyway for everything else.

IllusionMH commented on src/browser/public/Terminal.ts:
```diff
@@ -169,15 +169,16 @@ export class Terminal implements ITerminalApi {
+  public getOption(key: 'bellSound' | 'bellStyle' | 'cursorStyle' | 'fontFamily' | 'logLevel' | 'rendererType' | 'termName' | 'wordSeparator'): string;
   public getOption(key: 'allowTransparency' | 'cancelEvents' | 'convertEol' | 'cursorBlink' | 'disableStdin' | 'macOptionIsMeta' | 'rightClickSelectsWord' | 'popOnBell' | 'visualBell'): boolean;
   public getOption(key: 'cols' | 'fontSize' | 'letterSpacing' | 'lineHeight' | 'rows' | 'tabStopWidth' | 'scrollback'): number;
+  public getOption(key: 'fontWeight' | 'fontWeightBold'): string | number;
```
I've used wide type `string | number` to match previous wide `string`, however now `fontWeight` should always match `FontWeight` type because of custom normalization on `OptionsService` level.
Should I use narrow `FontWeight` type here and in `xterm.d.ts`?

Tyriar replied:
Yeah we can narrow to `FontWeight`, I think tsc ends up allowing anything right now anyway because of `getOption(key: any): any`. I wanted to move to something stricter later on which would fix this: https://github.com/xtermjs/xterm.js/issues/2522

IllusionMH commented on src/browser/public/Terminal.ts:
```diff
@@ -169,15 +169,16 @@ export class Terminal implements ITerminalApi {
   }
   public setOption(key: 'bellSound' | 'fontFamily' | 'termName' | 'wordSeparator', value: string): void;
-  public setOption(key: 'fontWeight' | 'fontWeightBold', value: 'normal' | 'bold' | '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900'): void;
+  public setOption(key: 'fontWeight' | 'fontWeightBold', value: 'normal' | 'bold' | '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900' | number): void;
```
Is there a reason why `value` uses union here instead of `FontWeight` type? Is it to improve intellisense/hovers or I should use narrow `FontWeight` type here and in `xterm.d.ts`?

Tyriar replied:
Yes, I think the reason was to make it more obvious using intellisense as API users wouldn't know what a `FontWeight` is within drilling into it. Let's stick to listing them out for now, again tsc is permissive because `setOption(key: any, value: any): void` is in the public API.

IllusionMH reviewed (commented):
Couple of questions

IllusionMH commented:
As per first feedback I've changed return type of `getOption` to `FontWeight` alias.
As for second comment about argument - I think there was no call to action so not changed.
Also added tests for this property behavior.

P.S. If function previously had return type `string` but now it also has `number` in return type - can this change be counted as minor/feature change or it's major/breaking change?
In fact previously it would return anything except falsy values, and now it's subset of actual previous return values, but superset of previously declared type.

Tyriar reviewed (approved):
> If function previously had return type string but now it also has number in return type - can this change be counted as minor/feature change or it's major/breaking change?

Yes it is technically a breaking change but I'm fine with letting this one go into a minor patch as it's such a small thing that maybe no one would eve. I don't know when v5 will happen and we don't want to block the VS Code change until then. I checked a few apps that use xterm.js and saw Hyper uses `getOption` only once for `bellSound`.
